### DISCUSSION TITLE
Why exceptions like ArgumentNullException do not have ParameterName when API consumer sets the "message"?

### DISCUSSION THREAD

**Discussion by Miljankg** [ardalis/GuardClauses#167] [Q&A] [answered]
Hi!

Thanks for the great lib! While using it, I noticed a small thing that was different than what I expected:
My Guard clause `Guard.Against.Null(someParameter, nameof(someParameter), "Some Parameter must not be null.");` throws ArgumentNullException when someParameter is null, as expected.

However, ArgumentNullException has a property `ParamName` that is in this case, null, and this is different than I expect. However, when I omit a custom message from my Guard clause, then the ArgumentNullException when thrown, does have ParamName set.

Going further, I found the code that does this, and it is indeed implemented like this (and it confirms that parameterName is omitted from the exception when the custom message is set):
File: `src/GuardClauses/GuardClauseExtensions.cs`
``` 
public static T Null<T>([JetBrainsNotNull] this IGuardClause guardClause, [NotNull, JetBrainsNotNull][ValidatedNotNull][JetBrainsNoEnumeration] T input, [JetBrainsNotNull][JetBrainsInvokerParameterName] string parameterName, string? message = null)
{
    if (input is null)
    {
        if (string.IsNullOrEmpty(message))
        {
            throw new ArgumentNullException(parameterName);
        }
        throw new ArgumentNullException(message, (Exception?)null);
    }

    return input;
}
```
This is probably very minor, but I would like to know if it is intentional and if it is, why?

Thanks in advance.

**Comment by ardalis** [comment, marked as answer]
Looks like a bug/omission. Do you want to add an issue or shall I?

**Reply by Miljankg** [reply]
I can do it no worries. Is it okay if I "fix" it and create a pull request?

**Reply by ardalis** [reply]
Of course! Make the issue first and have the PR say "Fixes #123"

**Reply by Miljankg** [reply]
Cool thanks! Will do as you described.
